### ISSUE TITLE
question on MNIST model 03_ConvBatchNorm

### ISSUE THREAD

**Issue by ignazwrobel007** [Microsoft/CNTK#905]
Hi,

I tested the model `03_ConvBatchNorm` (one of the MNIST examples) - it's impressive how fast it
delivers good results (error rate < 1 % after 3 epochs)
Starting with the given model I ran some tests with changed parameters
When trying different values for
  **minibatchSize** (default = 64)
I found that for very low values the training result gets worse.
A value of **1** delivers no useful result at all (**error rate > 50%**) - actually, the rate of training samples
per second is very bad at 1, so it might not make much sense; but I assume it should work anyway(?)
I also tested `02_Convolution`, minibatchSize = 1 works there (takes much longer, of course, but delivers
the same error rates as minibatchSize = 64)

So my question is: why is
  **minibatchSize = 1**
not working for `03_ConvBatchNorm`?

Please help

(test with GPU and also CPU_only)

**Comment by e-thereal** [comment]
Disclaimer: I've only read about batch normalization but I haven't used it myself.

tl;dr Batch normalization needs to calculate certain statistics of the input images in order to normalize features between different epochs without normalizing away useful information and requires a reasonably large sample size (=mini-batch size) in order to estimate those statistics reliably.

A potential problem for the training of deep networks is that during each update, both the feature extracting layers and the classification layers get updates. This makes it very difficult for the classification layers to learn a good mapping from features to output, because the extracted features constantly change. Batch normalization tries to alleviate this problem by normalizing the extracted features to a standard range, so that the classification layers can learn from a more stable feature representation. However, normalizing features is a difficult task, because you want them to be roughly the same (in magnitude) for the same image at different epoch, but you want them still to be different for different images in the same epoch. Or in other words, you don't want to normalize away the variability between images. In order to find a good compromise, normalization is always done on a batch of images instead of individual images. This way, batch normalization can estimate the variability between images within a batch and use that to regularize the normalization process. If the mini-batch size is too small, as in your case, batch normalization will underestimage the variability between images and essentially normalize useful variability way, which is why the training does not work.

The 02_Convolution network does not use batch normalization and does not need to calculate statistics for normalization. The mini-batch size only affects the number of updates per epoch, which has only a minor effect on the training process.

**Comment by ignazwrobel007** [comment]
thx, that was very helpful
Could you possibly provide recommendations for further reading regarding batch normalization?

**Comment by e-thereal** [comment]
You're welcom. The CNTK documentation itself provides some really good point, e.g., https://github.com/Microsoft/CNTK/wiki/Layers-Reference#batchnormalizationlayer-layernormalizationlayer-stabilizerlayer and references therein.

**Comment by cha-zhang** [comment]
Leave a comment
